sim/priest/priest.go

```diff
@@ -3,26 +3,297 @@ package priest
 import (
 	"time"
 
-	"github.com/wowsims/tbc/sim/core/proto"
 	"github.com/wowsims/tbc/sim/core"
+	"github.com/wowsims/tbc/sim/core/proto"
+	"github.com/wowsims/tbc/sim/core/stats"
 )
 
 type Priest struct {
 	core.Character
+	SelfBuffs
+	Talents proto.PriestTalents
+
+	// cached cast stuff
+	// TODO: aoe multi-target situations will need multiple spells ticking for each target.
+	MindFlaySpell        core.SimpleSpell
+	mindflayCastTemplate core.SimpleSpellTemplate
+
+	mindblastSpell        core.SimpleSpell
+	mindblastCastTemplate core.SimpleSpellTemplate
+
+	swdSpell        core.SimpleSpell
+	swdCastTemplate core.SimpleSpellTemplate
+
+	SWPSpell        core.SimpleSpell
+	swpCastTemplate core.SimpleSpellTemplate
+
+	VTSpell        core.SimpleSpell
+	vtCastTemplate core.SimpleSpellTemplate
+
+	ShadowfiendSpell    core.SimpleSpell
+	shadowfiendTemplate core.SimpleSpellTemplate
+
+	DevouringPlagueSpell    core.SimpleSpell
+	devouringPlagueTemplate core.SimpleSpellTemplate
+
+	// TODO: starshards
+}
+
+type SelfBuffs struct {
+	UseShadowfiend bool
 }
 
 func (priest *Priest) GetCharacter() *core.Character {
 	return &priest.Character
 }
 
 func (priest *Priest) AddRaidBuffs(raidBuffs *proto.RaidBuffs) {
-	raidBuffs.DivineSpirit = proto.TristateEffect_TristateEffectRegular
+	// TODO: handle this for raid sim
+	// raidBuffs.DivineSpirit = proto.TristateEffect_TristateEffectRegular
 }
+
 func (priest *Priest) AddPartyBuffs(partyBuffs *proto.PartyBuffs) {
+
+}
+
+func (priest *Priest) Init(sim *core.Simulation) {
+	priest.mindflayCastTemplate = priest.newMindflayTemplate(sim)
+	priest.mindblastCastTemplate = priest.newMindBlastTemplate(sim)
+	priest.swpCastTemplate = priest.newSWPTemplate(sim)
+	priest.vtCastTemplate = priest.newVTTemplate(sim)
+	priest.swdCastTemplate = priest.newSWDTemplate(sim)
+	priest.shadowfiendTemplate = priest.newShadowfiendTemplate(sim)
+	priest.devouringPlagueTemplate = priest.newDevouringPlagueTemplate(sim)
 }
 
-func (priest *Priest) Reset(sim *core.Simulation) {}
+func (priest *Priest) Reset(newsim *core.Simulation) {
+	// Cleanup and pending dots and casts
+	priest.mindblastSpell = core.SimpleSpell{}
+	priest.swdSpell = core.SimpleSpell{}
+	priest.MindFlaySpell = core.SimpleSpell{}
+	priest.SWPSpell = core.SimpleSpell{}
+	priest.VTSpell = core.SimpleSpell{}
+	priest.ShadowfiendSpell = core.SimpleSpell{}
+	priest.DevouringPlagueSpell = core.SimpleSpell{}
+}
+
+func (priest *Priest) Advance(sim *core.Simulation, elapsedTime time.Duration) {
+	// druid should never be outside the 5s window, use combat regen.
+	priest.Character.RegenManaCasting(sim, elapsedTime)
+}
 
 func (priest *Priest) Act(sim *core.Simulation) time.Duration {
-	return core.NeverExpires // makes the bot wait forever and do nothing.
+	return core.NeverExpires // does nothing
+}
+
+func (priest *Priest) applyTalentsToShadowSpell(cast *core.Cast, effect *core.SpellHitEffect) {
+	if cast.ActionID.SpellID == SpellIDSWD || cast.ActionID.SpellID == SpellIDMB {
```

Comment:
I really don't like having to check what type of spell this is. Should just put this logic in shadow_word_death.go and mind_blast.go directly.